be2-scala/src/test/scala/ch/epfl/pop/json/HighLevelProtocolSuite.scala

```diff
@@ -3,30 +3,122 @@ package ch.epfl.pop.json
+    // Add witnesses to the expected message
+    val witness = (WitnessSignaturePair(PublicKey(Base64Data("wit1")), Signature(Base64Data("sig1")))
+      :: WitnessSignaturePair(PublicKey(Base64Data("wit2")), Signature(Base64Data("sig2")))
+      :: Nil).reverse.foreach(w => {expected = expected.addWitnessSignature(w)})
+
+    val message = Message.buildFromJson(source)
+
+    message shouldBe a [Message]
+    message should equal (expected)
+  }
+
+  test("Parser: Empty json") {
+    val id: String = ""
+    val sender: String = ""
+    val signature: String = ""
+    val data: String = ""
+    val source: String =  s"""{
+                               |      "message_id": "$id",
+                               |      "sender": "$sender",
+                               |      "signature": "$signature",
+                               |      "data": "$data",
+                               |      "witness_signatures": []
+                               |    }""".stripMargin
 
-    val expected: Message = Message(
-      Base64Data("eyJjcmVhdGlvbiI6MTYzMTg4NzQ5NiwiaWQiOiJ4aWdzV0ZlUG1veGxkd2txMUt1b0wzT1ZhODl4amdYalRPZEJnSldjR1drPSIsIm5hbWUiOiJoZ2dnZ2dnIiwib3JnYW5pemVyIjoidG9fa2xaTHRpSFY0NDZGdjk4T0xOZE5taS1FUDVPYVR0YkJrb3RUWUxpYz0iLCJ3aXRuZXNzZXMiOltdLCJvYmplY3QiOiJsYW8iLCJhY3Rpb24iOiJjcmVhdGUifQ=="),
-      PublicKey(Base64Data("to_klZLtiHV446Fv98OLNdNmi-EP5OaTtbBkotTYLic=")),
-      Signature(Base64Data("2VDJCWg11eNPUvZOnvq5YhqqIKLBcik45n-6o87aUKefmiywagivzD4o_YmjWHzYcb9qg-OgDBZbBNWSUgJICA==")),
-      Hash(Base64Data("f1jTxH8TU2UGUBnikGU3wRTHjhOmIEQVmxZBK55QpsE=")),
-      WitnessSignaturePair(PublicKey(Base64Data("wit1")), Signature(Base64Data("sig1"))) :: WitnessSignaturePair(PublicKey(Base64Data("wit2")), Signature(Base64Data("sig2"))) :: Nil
-    )
 
+    var expected: Message = buildExpected(id, sender, signature, data)
     val message = Message.buildFromJson(source)
 
     message shouldBe a [Message]
     message should equal (expected)
   }
+
+  test("Parser: Empty witness array") {
+     val id: String = "f1jTxH8TU2UGUBnikGU3wRTHjhOmIEQVmxZBK55QpsE="
+    val sender: String ="to_klZLtiHV446Fv98OLNdNmi-EP5OaTtbBkotTYLic="
+    val signature: String = "2VDJCWg11eNPUvZOnvq5YhqqIKLBcik45n-6o87aUKefmiywagivzD4o_YmjWHzYcb9qg-OgDBZbBNWSUgJICA=="
+    val data: String = "eyJjcmVhdGlvbiI6MTYzMTg4NzQ5NiwiaWQiOiJ4aWdzV0ZlUG1veGxkd2txMUt1b0wzT1ZhODl4amdYalRPZEJnSldjR1drPSIsIm5hbWUiOiJoZ2dnZ2dnIiwib3JnYW5pemVyIjoidG9fa2xaTHRpSFY0NDZGdjk4T0xOZE5taS1FUDVPYVR0YkJrb3RUWUxpYz0iLCJ3aXRuZXNzZXMiOltdLCJvYmplY3QiOiJsYW8iLCJhY3Rpb24iOiJjcmVhdGUifQ=="
+    val source: String = s"""{
+                                 |      "message_id": "$id",
+                                 |      "sender": "$sender",
+                                 |      "signature": "$signature",
+                                 |      "data": "$data",
+                                 |      "witness_signatures": []
+                                 |    }""".stripMargin
```

Comment:
I see you are using this many times (i.e. formatting all fields of the message). Maybe create a function that would do it for you. This way, you won't have to change every single test if we decide (for example) to add a new field or rename `witness_signatures` to `witness_sig`.